### ISSUE TITLE
WML parser doesn't properly handle variables in non-quoted strings (GNA #16692)

### ISSUE THREAD

**Issue by wesnoth-bugs** [wesnoth/wesnoth#1236]
**[Original submission](https://gna.org/bugs/index.php?16692)** by anonymissimus on 2010-09-12

```

	[event]

		name=turn 2

		{DEBUG_MSG 1}

		{VARIABLE var 3}

		[event]

			name=turn $var

			delayed_variable_substitution=no

			{DEBUG_MSG 2}

		[/event]

	[/event]

```

The nested event's name expands to turn3 instead of turn 3, visible in the savegame and thus doesn't fire. It works if the space in turn $var is replaced by an underscore so that's a workaround.

------------------------------------------

lua.cpp, luaW_tovconfig, line 392: The config data about the nested event's name looks already suspicious when it comes from the lua interface and it looks like a bug that may have more than this effect.

(Reproduced on win xp)
Release: trunk r46431
Priority: 5 - Normal
Severity: 3 - Normal

**Comment by wesnoth-bugs** [comment]
**Modified on 2010-09-13**

**silene wrote:**

> Another workaround is to put quotes around "turn $var" so that the engine doesn't consider that the space is meaningless.

**Comment by wesnoth-bugs** [comment]
**Modified on 2010-12-05**

**silene changed status:** None -> Wont Fix

**silene closed the issue**

**Comment by wesnoth-bugs** [comment]
**Modified on 2011-02-10**

**sapient changed status:** Wont Fix -> Confirmed

**Comment by wesnoth-bugs** [comment]
**Modified on 2011-02-10**

**sapient wrote:**

> How is this "won't fix"? I consider this a valid bug.
> 
> 
> 
> The following code worked in 1.6 but no longer works:
> 
> 
> 
> 
> 
>     [event]
> 
>         name=prestart
> 
>         {VARIABLE ev_test 2}
> 
>     [/event]
> 
>     [event]
> 
>         name="turn $ev_test"
> 
>         first_time_only=no
> 
>         [message]
> 
>             speaker=narrator
> 
>             message=_"turn $|ev_test event has fired"
> 
>         [/message]
> 
>     [/event]
> 
> 
> 
> Note that the suggested workaround also did not work

**Comment by wesnoth-bugs** [comment]
**Modified on 2011-02-11**

**sapient reopened the issue**

**Comment by wesnoth-bugs** [comment]
**Modified on 2011-02-11**

**silene changed item group:** None of the others -> WML

**silene wrote:**

> As far as I can tell, the bug you report is completely unrelated to the original one from Anonymissimus, so I'm unassigning the issue from myself. The original bug was about the spacing being lost, which it isn't in your case, as can be seen by looking at the savefile.

**Comment by wesnoth-bugs** [comment]
**Modified on 2012-03-21**

**shadowmaster changed title:** name=turn $var - event\'s don\'t work -> Events with name=turn $var don't work unless quoted

**Comment by wesnoth-bugs** [comment]
**Modified on 2012-03-21**

**sapient wrote:**

> Regarding bug #16692, the test code in comment #2 still does not work in Wesnoth 1.10 (despite having quotes).
> 
> I suspect there is no variable substitution at all occurring in the event name.
> 
> 
> 
> Event names should perform variable substitution at the time that they are checked for a match with the fired event, not before. The contents of the [event] should be substituted at event firing time, unless documented otherwise; that is the default assumption, and the way it worked in previous versions.

**sapient changed title:** Events with name=turn $var don\'t work unless quoted -> Events with name=turn $var don't work

**Comment by Vultraz** [comment]
Still an issue almost 8 years later. Go figure :P

**Comment by Vultraz** [comment]
Ok, so did some debugging. This appears to be a problem either in the `config` class or the preprocessor itself. You can bypass all the `vconfig` and Lua stuff with a simpler test:

```
    [event]
        name=start
        {VARIABLE var 2}
    [/event]

    [event]
        name=turn $var

        [message]
            speaker=narrator
            message="turn 2 event fired"
        [/message]
    [/event]
```

I've tracked the bad key (`name=turn$var`) as far back as the `level` data passed to the `play_controller` constructor. Obviously, the space in the key value is lost somewhere in parsing, perhaps at the preprocessor level itself.

**Comment by Vultraz** [comment]
After more debugging, I've tracked down the cause of the bug, but I have no idea how to fix it. It's an issue with the WML parser, and I believe the relevant bit is this: https://github.com/wesnoth/wesnoth/blob/master/src/serialization/parser.cpp#L376-L379 (it's definitely an issue with the `parser::parse_variable()` function).

Essentially, the buggy `turn$var` string is constructed with three tokens: `turn`, `$`, and `var`. The first is of type `token::STRING`, so `previous_string` is true in the next loop. But then `$` is of type `token::MISC`, so it falls back to the default case and just appends it, so that's where we get `turn$`. If `$` were a string token, it would see that `previous_string` is true and properly append a space... but even when I change this line: https://github.com/wesnoth/wesnoth/blob/master/src/serialization/tokenizer.cpp#L140 to `if (is_alnum(current_) || current_ == '$') {`, it doesn't work.

So, in conclusion, I've tracked down the issue, but no idea how to fix it. Moving the `if(previous_string)` block to the default case or adding a fallthrough case for `token::MISC` to `token::STRING` both insert spaces in the middle of strings where they shouldn't be and break image paths.

Perhaps someone can see the logic I'm missing.
